File: frontend/src/api/cases.js
import client from "./client.js";

export const createCase = (name, description = null) =>
  client.post("/cases", { name, description }).then((r) => r.data);

export const listCases = (page = 1, pageSize = 20, status = null) => {
  const params = { page, page_size: pageSize };
  if (status) params.status = status;
  return client.get("/cases", { params }).then((r) => r.data);
};

export const getCase = (id) => client.get(`/cases/${id}`).then((r) => r.data);

export const updateCase = (id, data) =>
  client.put(`/cases/${id}`, data).then((r) => r.data);

export const deleteCase = (id) => client.delete(`/cases/${id}`);

export const exportCase = async (id) => {
  const response = await client.get(`/cases/${id}/export`, {
    responseType: "blob",
  });
  const url = URL.createObjectURL(response.data);
  const a = document.createElement("a");
  a.href = url;
  a.download = `case-${id}.zip`;
  a.click();
  URL.revokeObjectURL(url);
};

export const importCase = async (file) => {
  const formData = new FormData();
  formData.append("file", file);
  return client
    .post("/cases/import", formData, {
      headers: { "Content-Type": "multipart/form-data" },
    })
    .then((r) => r.data);
};

File: backend/app/routers/cases.py
import uuid

from fastapi import APIRouter, Depends, HTTPException, Query, UploadFile
from fastapi.responses import StreamingResponse
from sqlalchemy.ext.asyncio import AsyncSession

from app.dependencies import get_db
from app.schemas.case import CaseDetailOut, CaseIn, CaseListOut, CaseOut, CaseUpdate
from app.services import case_service

router = APIRouter()


@router.post("", response_model=CaseOut, status_code=201)
async def create_case(data: CaseIn, db: AsyncSession = Depends(get_db)):
    return await case_service.create_case(data, db)


@router.get("", response_model=CaseListOut)
async def list_cases(
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    status: str | None = Query(None),
    db: AsyncSession = Depends(get_db),
):
    return await case_service.list_cases(db, page=page, page_size=page_size, status=status)


@router.get("/{case_id}", response_model=CaseDetailOut)
async def get_case(case_id: uuid.UUID, db: AsyncSession = Depends(get_db)):
    case = await case_service.get_case(case_id, db)
    if case is None:
        raise HTTPException(status_code=404, detail="Case not found")
    return case


@router.put("/{case_id}", response_model=CaseOut)
async def update_case(
    case_id: uuid.UUID, data: CaseUpdate, db: AsyncSession = Depends(get_db)
):
    case = await case_service.update_case(case_id, data, db)
    if case is None:
        raise HTTPException(status_code=404, detail="Case not found")
    return case


@router.delete("/{case_id}", status_code=204)
async def delete_case(case_id: uuid.UUID, db: AsyncSession = Depends(get_db)):
    deleted = await case_service.delete_case(case_id, db)
    if not deleted:
        raise HTTPException(status_code=404, detail="Case not found")


@router.get("/{case_id}/export")
async def export_case(case_id: uuid.UUID, db: AsyncSession = Depends(get_db)):
    zip_buffer = await case_service.export_case(case_id, db)
    if zip_buffer is None:
        raise HTTPException(status_code=404, detail="Case not found")
    return StreamingResponse(
        zip_buffer,
        media_type="application/zip",
        headers={"Content-Disposition": f"attachment; filename=case-{case_id}.zip"},
    )


@router.post("/import", response_model=CaseOut, status_code=201)
async def import_case(file: UploadFile, db: AsyncSession = Depends(get_db)):
    raw = await file.read()
    try:
        return await case_service.import_case(raw, db)
    except Exception as exc:
        raise HTTPException(status_code=400, detail=f"Import failed: {exc}")
